src/Controllers/FilesController.js

```diff
@@ -31,12 +31,14 @@ export class FilesController extends AdaptableController {
     }
 
     const location = this.adapter.getFileLocation(config, filename);
```

Comment:
Don't we have a problem here deciding the location prior to your hooks being called?

Reply:
Yes, this is the same problem as described here: https://github.com/parse-community/parse-server/issues/6518 the location is computed before the triggers and adapters get a chance to potentially modify the file name.

Reply:
There is currently no way to change the name of the Parse.File because there is no setter. I believe with these hooks you could create a new Parse.File (with a new name) using the same data as the original file and return it in the hook. Your new file (with the new file name) will be saved instead.

Reply:
@stevestencil Would it be possible to add a test case for this. When a client SDK asks to save file A.txt and in beforeSave you create new File with name B.txt and the same contents, verify that it is actually saved into B.txt? That would be awesome to consider this practice as a blessed way to change filename during the upload. To explain some background. Client SDK wants to save A.txt but when pushing to S3 you want to make the filename  YYYY-MM-DD-A.txt so that you can later clean up files on S3.

Reply:
I wrote a test case for your example and it passes. You'll just have to make sure preserveFileName is set to true in your parse-server config so that the random text does not get prepended to your file names.

Reply:
Sorry to slow this down, I can't wait to get it in and use it in production, but what happens when ``preserveFileName: false`` which is the default. The beforeSave handler should receive a filename passed from the client sdk prepended with a random string, and will have a change to work from that, correct? It should work both ways.

Reply:
In other words, prepending a filename passed in from client SDK is fine IMHO, this is the default server behavior to make sure that all saved files are unique. The beforeSave trigged should get called on such file, and there if you change anything, it should still work, regardless whether the filename in beforeSave was kept intact from client SDL (``preserveFilename: true``) or whether it was prepended with random chars (``preserveFilename: false`` - the default behavior). Correct?

Reply:
no matter if preserveFileName is true or false, the beforeSaveFile hook will get the file name as it came from the client (lets say foo.txt)... The file name gets changed by adding the random text to the beginning of the file name inside of parse-server. if you set preserveFileName to false this will not happen.

So in a nutshell. The random text gets added after the beforeSaveFile hook and before the file actually saves to whatever storage solution you're using.